### ISSUE TITLE
'Missing return statement' error when AsyncExitStack is used

### ISSUE THREAD

**Issue by asvetlov** [python/mypy#15715]
**Bug Report**

When a function use `AsyncExitStack` mypy blames that return value is missing (but it is present).

Replacement `AsyncExitStack` with plain async context manager doesn't generate a bug.

**To Reproduce**

```python
from contextlib import AsyncExitStack

async def f() -> int:
    return 1

async def g() -> int:
    async with AsyncExitStack():
        return await f()
```

[Snippet on mypy-play.net](https://mypy-play.net/?mypy=latest&python=3.11&gist=9ee98d6c63cadc8c92ac7a041959ead2)

**Expected Behavior**

The return statement is present, I don't expect any errors.

**Actual Behavior**
```
...: error: Missing return statement  [return]
Found 1 error in 1 file (checked 1 source file)
```

**Your Environment**

The bug is reproducible with plain mypy, nothing special.

- Mypy version used: 1.4.1
- Mypy command-line flags: none
- Mypy configuration options from `mypy.ini` (and other config files): none
- Python version used: 3.11

**Comment by AlexWaygood** [comment]
Hi @asvetlov! Unfortunately this isn't a false positive error -- though I definitely agree that it can sometimes be annoying, and I've had to work around it in some of my own projects too. I gave an in-depth explanation about why mypy gives this error here: https://github.com/python/typeshed/issues/10050#issuecomment-1510019445.

**Comment by asvetlov** [comment]
Thank you for the explanation.

Wild idea.
Maybe (Async)ExitStack can accept a type argument which marks that none of inner context managers swallow exceptions?

E.g. `ExitStack` by default keeps the current behavior but `ExitStack[True]` tells to type checker that the stack doesn't swallow?

https://peps.python.org/pep-0696/ can be used for omitting generic arg by default.

**Comment by AlexWaygood** [comment]
> Maybe (Async)ExitStack can accept a type argument which marks that none of inner context managers swallow exceptions?

I think it would _probably_ be too disruptive unless and until PEP 696 is accepted by the Steering Council and implemented by mypy. But you're welcome to file an experimental PR over at typeshed and see what mypy_primer says :)



> E.g. `ExitStack` by default keeps the current behavior but `ExitStack[True]` tells to type checker that the stack doesn't swallow?

(It would need to be `Literal[]`, since `True` is a value, not a type. And it would need to be `ExitStack[Literal[False]]` to indicate that exceptions aren't swallowed — it's `__(a)exit__` methods with truthy return values that cause exceptions to be swallowed at runtime.)

**Comment by AlexWaygood** [comment]
It's definitely not mypy's most user-friendly error message! There's some ongoing discussion in a discuss.python.org thread (that I can't find right now on my mobile) about possible solutions, but it's not an easy thing to fix :/

**Comment by jenstroeger** [comment]
Thank you @AlexWaygood! I poked aronud the Python discussions and found two or three rather lengthy threads on the ExitStack and on missing return statements. If you happen to find the conversation you mentioned above, I’d be keen to read through it 🤓
